stor/tests/test_s3.py

```diff
@@ -1085,19 +1108,21 @@ def test_upload_w_use_manifest_single_file(self, mock_getsize, mock_files):
             S3Path('s3://bucket/path').upload(['file'],
                                               use_manifest=True)
 
-    @mock.patch('stor.s3.ThreadPool', autospec=True)
-    def test_upload_object_threads(self, mock_pool, mock_getsize, mock_files):
+    @mock.patch("stor.s3.as_completed", return_value=[MockFuture() for x in range(0, 20)])
+    @mock.patch("stor.s3.ThreadPoolExecutor", return_value=MockExecutor())
+    def test_upload_object_threads(
+        self, mock_pool, mock_completed, mock_getsize, mock_files
+    ):
         mock_files.return_value = {
             'file%s' % i: 20
             for i in range(20)
         }
         mock_getsize.return_value = 20
-        mock_pool.return_value.imap_unordered.return_value.next.side_effect = StopIteration
 
         s3_p = S3Path('s3://bucket')
         with settings.use({'s3:upload': {'object_threads': 20}}):
             s3_p.upload(['test'])
-        mock_pool.assert_called_once_with(20)
+        mock_pool.assert_called_once_with(max_workers=20)
```

Comment:
this test was previously pretty simple just asserting that the mock pool was called with the correct args -> can we do something like the following instead?

```
# either in decorator or context manager -> mock object and spec
mock_pool = mock.patch("stor.s3.ThreadPoolExecutor", autospec=True)
mock_executor = mock.patch("stor.s3. as_completed", autospec=True)

# return a function which takes a callable and val and returns a val
# or can just hard code .return_value = 5 or something
mock_pool.return_value.submit.side_effect = lambda fun, val: val 

# return a list of the length of input - can be anything with .result() function
mock_as_completed.side_effect = lambda futures: [MockFuture for future in futures]


# and assert the mocks were called with right args etc.
```

Reply:
Based on how the submit mocking was going, I left it in as calling the mock class. Added additional asserts for as_completed being called correctly in [834443b](https://github.com/counsyl/stor/pull/144/commits/834443bc0df6f9ce9d0718372cbd81fc865e7259).